sdk/storage/azure-storage-blob/azure/storage/blob/_shared/avro/datafile.py

```diff
@@ -0,0 +1,301 @@
+# Exceptions
+
+
+class DataFileException(schema.AvroException):
+  """Problem reading or writing file object containers."""
+
+  def __init__(self, msg):
+    super(DataFileException, self).__init__(msg)
+
+# ------------------------------------------------------------------------------
+
+
+class DataFileReader(object):
+  """Read files written by DataFileWriter."""
+
+  # TODO: allow user to specify expected schema?
+  # TODO: allow user to specify the encoder
+  def __init__(self, reader, datum_reader):
+    """Initializes a new data file reader.
+
+    Args:
+      reader: Open file to read from.
+      datum_reader: Avro datum reader.
+    """
+    self._reader = reader
+    self._raw_decoder = avro_io.BinaryDecoder(reader)
+    self._datum_decoder = None # Maybe reset at every block.
+    self._datum_reader = datum_reader
+
+    # read the header: magic, meta, sync
+    self._read_header()
+
+    # ensure codec is valid
+    avro_codec_raw = self.GetMeta('avro.codec')
+    if avro_codec_raw is None:
+      self.codec = "null"
+    else:
+      self.codec = avro_codec_raw.decode('utf-8')
+    if self.codec not in VALID_CODECS:
+      raise DataFileException('Unknown codec: %s.' % self.codec)
+
+    self._file_length = self._GetInputFileLength()
+
+    # get ready to read
+    self._block_count = 0
+    self.datum_reader.writer_schema = (
+        schema.parse(self.GetMeta(SCHEMA_KEY).decode('utf-8')))
+
+  def __enter__(self):
+    return self
+
+  def __exit__(self, type, value, traceback):
+    # Perform a close if there's no exception
+    if type is None:
+      self.close()
+
+  def __iter__(self):
+    return self
+
+  def __next__(self):
+    """Implements the iterator interface."""
+    return next(self)
+
+  # read-only properties
+  @property
+  def reader(self):
+    return self._reader
+
+  @property
+  def raw_decoder(self):
+    return self._raw_decoder
+
+  @property
+  def datum_decoder(self):
+    return self._datum_decoder
+
+  @property
+  def datum_reader(self):
+    return self._datum_reader
+
+  @property
+  def sync_marker(self):
+    return self._sync_marker
+
+  @property
+  def meta(self):
+    return self._meta
+
+  @property
+  def file_length(self):
+    """Length of the input file, in bytes."""
+    return self._file_length
+
+  # read/write properties
+  @property
+  def block_count(self):
+    return self._block_count
+
+  def GetMeta(self, key):
+    """Reports the value of a given metadata key.
+
+    Args:
+      key: Metadata key (string) to report the value of.
+    Returns:
+      Value associated to the metadata key, as bytes.
+    """
+    return self._meta.get(key)
+
+  def SetMeta(self, key, value):
+    """Sets a metadata.
+
+    Args:
+      key: Metadata key (string) to set.
+      value: Metadata value to set, as bytes.
+    """
+    if isinstance(value, str):
+      value = value.encode('utf-8')
+    self._meta[key] = value
+
+  def _GetInputFileLength(self):
```

Comment:
I'm not sure if this will work with Quick Query, since QQ streams don't have a defined length.

Reply:
I managed to get rid of that. Looks like stream-like object starts to return empty when end is reached.